src/utils/network.ts

```diff
@@ -284,13 +297,35 @@ export const getOpenPorts = async (network: Network): Promise<OpenPorts | undefi
   // filter out nodes that are already started since their ports are in use by themselves
   const bitcoin = network.nodes.bitcoin.filter(n => n.status !== Status.Started);
   if (bitcoin.length) {
-    const existingPorts = bitcoin.map(n => n.ports.rpc);
-    const openPorts = await getOpenPortRange(existingPorts);
+    let existingPorts = bitcoin.map(n => n.ports.rpc);
+    let openPorts = await getOpenPortRange(existingPorts);
     if (openPorts.join() !== existingPorts.join()) {
       openPorts.forEach((port, index) => {
         ports[bitcoin[index].name] = { rpc: port };
       });
     }
+
+    existingPorts = bitcoin.map(n => n.ports.zmqpubrawblock);
+    openPorts = await getOpenPortRange(existingPorts);
+    if (openPorts.join() !== existingPorts.join()) {
+      openPorts.forEach((port, index) => {
+        ports[bitcoin[index].name] = {
+          ...(ports[bitcoin[index].name] || {}),
```

Comment:
Sorry, I'm not sure I understand, isn't this covered by https://github.com/jamaljsr/polar/blob/34da4f7d110f198a2407ae2a8116c2b2bbd99527/src/utils/network.spec.ts#L66

The code here is the same as the one for lnd ports.

Reply:
The test you linked to does not cover the `|| {}` code path. This test doesn't even make it to the `openPorts.forEach(...)` statement because the the `detect-port` mock always returns the same ports unchanged and therefore `openPorts.join() !== existingPorts.join()` is false. This is why the return value is undefined.

To cover this code path, you have to write a test where `ports[bitcoin[index].name]` is still undefined when it gets to line 307. The only way to induce this condition is to make sure the REST port is unchanged so that is does not set the value of `ports[bitcoin[index].name]` on line 298. The simplest way to do this is to make separate tests for each port similar to how the lightning REST/GRPC tests are separate.

FYI: You can run the tests with coverage by running `yarn test:ci`. It will display which lines are uncovered in the console and will also generate an html report in `coverage/lcov-report/index.html` which is really helpful.

Reply:
@jamaljsr 
> The test you linked to does not cover the || {} code path. This test doesn't even make it to the openPorts.forEach(...) statement because the the detect-port mock always returns the same ports unchanged and therefore openPorts.join() !== existingPorts.join() is false. This is why the return value is undefined.

You're right I was confused.

> To cover this code path, you have to write a test where ports[bitcoin[index].name] is still undefined when it gets to line 307. The only way to induce this condition is to make sure the REST port is unchanged so that is does not set the value of ports[bitcoin[index].name] on line 298. The simplest way to do this is to make separate tests for each port similar to how the lightning REST/GRPC tests are separate.

Gotcha! I'll have an update ready sometime today.

> FYI: You can run the tests with coverage by running yarn test:ci. It will display which lines are uncovered in the console and will also generate an html report in coverage/lcov-report/index.html which is really helpful.

Yes, that is really helpful actually.

Reply:
Fixed.